### DISCUSSION TITLE
Sending a server message to all clients

### DISCUSSION THREAD

**Discussion by AltAccountForNobody** [FabricMC/fabric-api#3536] [Mod Dev Support]
### Description

I've been stumped attempting to send a message to all clients connected to a server in my mod because I can't figure out how to access the list of players in a normal function. Is there a way to send a message to all players connected to a server with the mod installed without access to the MinecraftServer instance, or is there a way to obtain the MinecraftServer instance? Any help is appreciated!

**Comment by haykam821** [comment]
If you need access to the current `MinecraftServer` instance, you can capture it and store it in a field using the server lifecycle events provided by Fabric API:

```java
ServerLifecycleEvents.SERVER_STARTING.register(server -> {
	MyMod.server = server;
});

ServerLifecycleEvents.SERVER_STOPPING.register(server -> {
	MyMod.server = null;
});
```

Elsewhere, you can reference `MyMod.server` (with a null check to ensure the server actually exists!) and perform any action, including sending messages to all connected players:

```java
if (MyMod.server != null) {
	for (ServerPlayerEntity player : MyMod.server.getPlayerManager().getPlayerList()) {
		player.sendMessage(Text.literal("Hello, world!"));
	}
}
```
